<a href="https://colab.research.google.com/github/BrunoDkno368/Limpieza-y-Normalizaci-n-de-datos-ETL/blob/main/LimpiezayNormalizacion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#EXTRAT (E)

In [ ]:
import pandas as pd
import numpy as np


In [ ]:
df_raw = pd.read_csv('/content/drive/MyDrive/ETL/Datos sucios/dataset_sucio.csv')

Copiamos el df para no modificar el df real

In [ ]:
df = df_raw.copy()

#TRANSFORM(T)

##Info del data set

In [ ]:
print(' La cantidad de filas es:', df.shape[0], '\n', 'La cantidad de columnas es:', df.shape[1] )

 La cantidad de filas es: 11 
 La cantidad de columnas es: 6


In [ ]:
df

id  nombre     edad    fecha_registro monto_compra     provincia
0    1    Juan       25        2024-01-05       1500.5       Tucumán
1    2   María  treinta        05/01/2024     1.200,30       tucuman
2    2  Pedro        30        2024/01/05          NaN      Tucuman 
3    3     Ana       -5      enero 5 2024          500  Buenos aires
4    4     NaN       40               NaN         -300           NaN
5    5               22        2024-13-01          NaN         Bs As
6    6    Luis      NaN        2024-01-32      9999999          CABA
7    7   Luisa       55  2024-01-05 12:00         1000         caba 
8    8    Ju@n      200               NaN                    Tucumán
9    9     NaN       33          2024-1-5       2500.0       TUCUMÁN
10  10  Mar ia      NaN        2024-01-05          100       mendoza

##Normalizacion columas que contienen nombre

In [ ]:
for col in ['nombre', 'provincia']:
  df[col] = df[col].astype(str).str.strip()
  df[col] = df[col].replace({'nan': np.nan})

df['provincia'] = df['provincia'].str.title()

In [ ]:
df[['nombre','provincia']]

nombre     provincia
0     Juan       Tucumán
1    María       Tucuman
2    Pedro       Tucuman
3      Ana  Buenos Aires
4      NaN           NaN
5                  Bs As
6     Luis          Caba
7    Luisa          Caba
8     Ju@n       Tucumán
9      NaN       Tucumán
10  Mar ia       Mendoza

In [ ]:
df['provincia'].unique()

array(['Tucumán', 'Tucuman', 'Buenos Aires', nan, 'Bs As', 'Caba',
       'Mendoza'], dtype=object)

vemos que tucuman esta con acento y sin acento pero es la misma provincia y en Bs As es Buenos aires

In [ ]:
df['provincia']= df['provincia'].replace({'Tucuman':'Tucumán'})
df['provincia'] = df['provincia'].replace({'Bs As':'Buenos Aires'})
df['provincia'].unique()

array(['Tucumán', 'Buenos Aires', nan, 'Caba', 'Mendoza'], dtype=object)

In [ ]:
df

id  nombre     edad    fecha_registro monto_compra     provincia
0    1    Juan       25        2024-01-05       1500.5       Tucumán
1    2   María  treinta        05/01/2024     1.200,30       Tucumán
2    2   Pedro       30        2024/01/05          NaN       Tucumán
3    3     Ana       -5      enero 5 2024          500  Buenos Aires
4    4     NaN       40               NaN         -300           NaN
5    5               22        2024-13-01          NaN  Buenos Aires
6    6    Luis      NaN        2024-01-32      9999999          Caba
7    7   Luisa       55  2024-01-05 12:00         1000          Caba
8    8    Ju@n      200               NaN                    Tucumán
9    9     NaN       33          2024-1-5       2500.0       Tucumán
10  10  Mar ia      NaN        2024-01-05          100       Mendoza

Verificamos la columna edad

##Columna EDAD

In [ ]:
df['edad'].unique()

array(['25', 'treinta', '30', '-5', '40', '22', nan, '55', '200', '33'],
      dtype=object)

In [ ]:
df['edad']= df['edad'].replace({'treinta': 30})

vemos que hay varias edades que estan mal -5, 200 y escrito en numeros

In [ ]:
df['edad'] = pd.to_numeric(df['edad'], errors= 'coerce')
df.loc[(df["edad"] < 0) | (df["edad"] > 120), "edad"] = np.nan

In [ ]:
df['edad'] = df['edad'].fillna(round(df['edad'].mean()))

In [ ]:
df['edad'] = df['edad'].astype(int)

In [ ]:
df

id  nombre  edad    fecha_registro monto_compra     provincia
0    1    Juan    25        2024-01-05       1500.5       Tucumán
1    2   María    30        05/01/2024     1.200,30       Tucumán
2    2   Pedro    30        2024/01/05          NaN       Tucumán
3    3     Ana    34      enero 5 2024          500  Buenos Aires
4    4     NaN    40               NaN         -300           NaN
5    5            22        2024-13-01          NaN  Buenos Aires
6    6    Luis    34        2024-01-32      9999999          Caba
7    7   Luisa    55  2024-01-05 12:00         1000          Caba
8    8    Ju@n    34               NaN                    Tucumán
9    9     NaN    33          2024-1-5       2500.0       Tucumán
10  10  Mar ia    34        2024-01-05          100       Mendoza

## Columna NOMBRE

In [ ]:
df['nombre'].unique()

array(['Juan', 'María', 'Pedro', 'Ana', nan, '', 'Luis', 'Luisa', 'Ju@n',
       'Mar ia'], dtype=object)

Verificamos varios errores en los datos de la columna nombre
'María' y 'Maria' tambien esta 'Mar ia'. otro error que vemos
es 'Juan' y 'Ju@an'. Tambien encontamos nan

Normalizacion basicas, elim de espacios adelante, atras y simbolos especiales

In [ ]:
df['Nombre_limpio'] = df['nombre']

In [ ]:
df['Nombre_limpio'] = (df['Nombre_limpio'].
                       astype(str).                           #tipo de datos que sea string
                       str.strip().                           #eliminamos espacios al comienzo y al final
                       str.replace(r"\s+", " ", regex=True))  #eliminamos simbolos especiales


In [ ]:
df["Nombre_limpio"] = df["Nombre_limpio"].replace({"": np.nan, "nan": np.nan})

In [ ]:
print(display(df[['Nombre_limpio','nombre']]))

Nombre_limpio  nombre
0           Juan    Juan
1          María   María
2          Pedro   Pedro
3            Ana     Ana
4            NaN     NaN
5            NaN        
6           Luis    Luis
7          Luisa   Luisa
8           Ju@n    Ju@n
9            NaN     NaN
10        Mar ia  Mar ia

None


Vemos varios errores, 2 registros NaN, Ju@n, Mar ia. vamos a corregir los errores tratando de no perder ninguno ya que son pocos registros

Nan lo reemplazamos por Sin datos

In [ ]:
df['Nombre_limpio'].fillna('Sin datos', inplace=True)

/tmp/ipython-input-962945600.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Nombre_limpio'].fillna('Sin datos', inplace=True)


Ju@n reemplazamos por Juan

primero limpiamos los simbolos que no corresponden

In [ ]:
df["Nombre_limpio"] = df["Nombre_limpio"].str.replace(
    r"[^A-Za-zÁÉÍÓÚáéíóúÑñ ]",
    "",
    regex=True
)

paso de Ju@n a Jun aun tenemos que reeplazar para que quede unificado

In [ ]:
df['Nombre_limpio'] = df['Nombre_limpio'].replace({'Jun':'Juan'})

queda totalemente limpio Ju@n en Juan

pasamos de Mar ia a María

In [ ]:
df['Nombre_limpio'] = df['Nombre_limpio'].replace({'Mar ia':'María'})

queda remplazado el dato Mar ia

In [ ]:
df['Nombre_limpio'].unique()

array(['Juan', 'María', 'Pedro', 'Ana', 'Sin datos', 'Luis', 'Luisa'],
      dtype=object)

In [ ]:
df['nombre']=df['Nombre_limpio']

In [ ]:
df.drop(['Nombre_limpio'], axis=1, inplace=True)

In [ ]:
df.columns

Index(['id', 'nombre', 'edad', 'fecha_registro', 'monto_compra', 'provincia'], dtype='object')

##Columna PROVINCIA

In [ ]:
df['provincia'].unique()

array(['Tucumán', 'Buenos Aires', nan, 'Caba', 'Mendoza'], dtype=object)

En el campo Provincia hay 1 dato Nan lo pasamos a Sin datos

In [ ]:
df['provincia'].fillna('Sin datos', inplace=True)

/tmp/ipython-input-2872165264.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['provincia'].fillna('Sin datos', inplace=True)


##Columna MONTO DE COMPRA

Verificamos el typo de la columa

In [ ]:
df['monto_compra'].dtypes

dtype('O')

In [ ]:
df['monto_compra'].unique()

array(['1500.5', '1.200,30', nan, '500', '-300', '9999999', '1000', ' ',
       '2500.0', '100'], dtype=object)

In [ ]:
df['monto_limpio'] = (df['monto_compra']
                      .astype(str)                  #Pasamos a string los datos de la columna seleccionada
                      .str.strip())                   #Eliminamos los datos al comienzo y al final


pasamos a String ya que como los dato estan sucios no estan en el mismo formato lo unificamos para poder trabajarlo mejor

limpiamos las columnas vacias y las columas nan en NaN

In [ ]:
print(display(df[['monto_limpio', 'monto_compra']]))

monto_limpio monto_compra
0        1500.5       1500.5
1      1.200,30     1.200,30
2           nan          NaN
3           500          500
4          -300         -300
5           nan          NaN
6       9999999      9999999
7          1000         1000
8                           
9        2500.0       2500.0
10          100          100

None


Tenemos un numero que es 1.200,30 tenemos que logrrar pasarlo 1200.30 si no cuando lo pasemos a numerico va a dar NaN

In [ ]:
df["monto_limpio"] = (
    df["monto_limpio"]
    .astype(str)
    .str.replace(".", "", regex=False)
    .str.replace(",", ".", regex=False)
)

df["monto_limpio"] = pd.to_numeric(df["monto_limpio"], errors="coerce")

Como ya tenemos todos los dato en numericos o Nan vamos a pasarlo a numerico

In [ ]:
df['monto_limpio'] = pd.to_numeric(df['monto_limpio'], errors= 'coerce')

Vemos que hay un monto que es -300 y un monto 9999999 ya que vemos un saldo negativo y uno muy grande vamos a pasarlo a NaN ya que por regla de negocio no podemos tener una compra negativa y no lopodemos poner como positivo porque puede modificar la calidad de los datos

In [ ]:
df.loc[df["monto_limpio"] < 0, "monto_limpio"] = np.nan
df.loc[df["monto_limpio"] > 90000, "monto_limpio"] = np.nan

In [ ]:
print(display(df[['monto_limpio', 'monto_compra']]))

monto_limpio monto_compra
0        15005.0       1500.5
1         1200.3     1.200,30
2            NaN          NaN
3          500.0          500
4            NaN         -300
5            NaN          NaN
6            NaN      9999999
7         1000.0         1000
8            NaN             
9        25000.0       2500.0
10         100.0          100

None


como tenemos varios Nan vamos a redondear con los datos que nos quedan

In [ ]:
df['monto_limpio'].fillna(df['monto_limpio'].mean().round(0), inplace=True)

/tmp/ipython-input-2660984751.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['monto_limpio'].fillna(df['monto_limpio'].mean().round(0), inplace=True)


con esto concluimos la limpieza de la columna monto de compra

In [ ]:
print(display(df[['monto_limpio', 'monto_compra']]))

monto_limpio monto_compra
0        15005.0       1500.5
1         1200.3     1.200,30
2         7134.0          NaN
3          500.0          500
4         7134.0         -300
5         7134.0          NaN
6         7134.0      9999999
7         1000.0         1000
8         7134.0             
9        25000.0       2500.0
10         100.0          100

None


asignamos monto limpio a la columna original de monto de compra

In [ ]:
df['monto_compra']=df['monto_limpio']

In [ ]:
df.drop(['monto_limpio'], axis=1, inplace=True)

In [ ]:
df.columns

Index(['id', 'nombre', 'edad', 'fecha_registro', 'monto_compra', 'provincia'], dtype='object')

##Columna FECHA DE REGISTRO

In [ ]:
df['fecha_registro'].unique()

array(['2024-01-05', '05/01/2024', '2024/01/05', 'enero 5 2024', nan,
       '2024-13-01', '2024-01-32', '2024-01-05 12:00', '2024-1-5'],
      dtype=object)

Vemos varios errores y diferentes formatos 2024-01-05, enero 5 2024, 2024-13-01, 2024-01-05 12:00

Normalizamos todo a texto

In [ ]:
df['fecha_str'] = df['fecha_registro'].astype(str)

In [ ]:

meses = {
    'enero': 'january',
    'febrero': 'february',
    'marzo': 'march',
    'abril': 'april',
    'mayo': 'may',
    'junio': 'june',
    'julio': 'july',
    'agosto': 'august',
    'septiembre': 'september',
    'octubre': 'october',
    'noviembre': 'november',
    'diciembre': 'december'
}

for es, en in meses.items():
    df['fecha_str'] = df['fecha_str'].str.replace(es, en, regex=False)


In [ ]:
df['fecha_str']

0           2024-01-05
1           05/01/2024
2           2024/01/05
3       january 5 2024
4                  nan
5           2024-13-01
6           2024-01-32
7     2024-01-05 12:00
8                  nan
9             2024-1-5
10          2024-01-05
Name: fecha_str, dtype: object

In [ ]:
df['fecha_str'] = df['fecha_str'].str.replace('/','-',regex=False)

In [ ]:
df['fecha_str']

0           2024-01-05
1           05-01-2024
2           2024-01-05
3       january 5 2024
4                  nan
5           2024-13-01
6           2024-01-32
7     2024-01-05 12:00
8                  nan
9             2024-1-5
10          2024-01-05
Name: fecha_str, dtype: object

In [ ]:
df['fecha_str'] = df['fecha_str'].replace({'2024-01-05 12:00': '2024-01-05',
                                           '05-01-2024': '2024-01-05',
                                          'january 5 2024':'2024-01-05',
                                          '2024-13-01':'2024-01-13',
                                          '2024-1-5':'2024-01-05'})

In [ ]:
df['fecha_str']

0     2024-01-05
1     2024-01-05
2     2024-01-05
3     2024-01-05
4            nan
5     2024-01-13
6     2024-01-32
7     2024-01-05
8            nan
9     2024-01-05
10    2024-01-05
Name: fecha_str, dtype: object

In [ ]:
#df['fecha_str']

In [ ]:
df['fecha_str'] = pd.to_datetime(df['fecha_str'], format='%Y-%m-%d', errors='coerce')

In [ ]:
display(df[['fecha_str', 'fecha_registro']])

fecha_str    fecha_registro
0  2024-01-05        2024-01-05
1  2024-01-05        05/01/2024
2  2024-01-05        2024/01/05
3  2024-01-05      enero 5 2024
4         NaT               NaN
5  2024-01-13        2024-13-01
6         NaT        2024-01-32
7  2024-01-05  2024-01-05 12:00
8         NaT               NaN
9  2024-01-05          2024-1-5
10 2024-01-05        2024-01-05

In [ ]:
df['fecha_registro']= df['fecha_str']

In [ ]:
df.drop(['fecha_str'],axis=1, inplace=True)

De esta manera queda limpio los registros de cada columna

In [ ]:
df

id     nombre  edad fecha_registro  monto_compra     provincia
0    1       Juan    25     2024-01-05       15005.0       Tucumán
1    2      María    30     2024-01-05        1200.3       Tucumán
2    2      Pedro    30     2024-01-05        7134.0       Tucumán
3    3        Ana    34     2024-01-05         500.0  Buenos Aires
4    4  Sin datos    40            NaT        7134.0     Sin datos
5    5  Sin datos    22     2024-01-13        7134.0  Buenos Aires
6    6       Luis    34            NaT        7134.0          Caba
7    7      Luisa    55     2024-01-05        1000.0          Caba
8    8       Juan    34            NaT        7134.0       Tucumán
9    9  Sin datos    33     2024-01-05       25000.0       Tucumán
10  10      María    34     2024-01-05         100.0       Mendoza

In [ ]:
df.to_csv('datos_limpios.csv', index=False)